File: JavaScript/Arrays/Scaler/max_sum_contiguous_subarray.js
function solve(A) {
  let maxSum = A[0]; // 5
  let currSum = A[0]; // 2

  for (let i = 1; i < A.length; i++) {
    currSum = Math.max(currSum + A[i], A[i]);
    maxSum = Math.max(maxSum, currSum);
  }

  return maxSum;
}

console.log(solve([-2, 1, -3, 4, -1, 2, 1, -5, 4]));

File: python/Arrays/Scaler/max_sum_contiguous_subarray.py
def solve(A):
    n = len(A)
    curr_sum = A[0]
    max_sum = A[0]

    for i in range(len(A)):
        curr_sum = max(curr_sum + A[i], A[i])
        max_sum = max(max_sum, curr_sum)
    return max_sum


print(solve([-2, 1, -3, 4, -1, 2, 1, -5, 4]))
# TC:O(N)
# SC: O(1)
